### PR TITLE
Feature/search result

### PR DESCRIPTION
**Pull request by ClaudivanFilho**
#### What is the purpose of this pull request?

- App Gallery was transformed into SearchResult.

#### Screenshots or example usage
screenshot from 2018-06-08 09-56-05

[workspace link](https://claudivan--storecomponents.myvtex.com/Eletronics?map=c&orderBy=OrderByTopSaleDESC&page=1)

#### Types of changes

* [ ] Bug fix (a non-breaking change which fixes an issue)
* [x] New feature (a non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: manifest.json [hunk 1]
```diff
@@ -1,9 +1,9 @@
 {
-  "name": "gallery",
+  "name": "search-result",
   "vendor": "vtex",
-  "version": "0.2.0",
-  "title": "VTEX Gallery",
-  "description": "A gallery component",
+  "version": "0.1.0",
```
Comment (andreldsa) [inline manifest.json]: Why downgrade de version? This still is the gallery.

File: react/SearchResult.js [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+    const { facetsQuery: { facets }, searchQuery } = this.props
+    const query = searchQuery.variables.query
+    const map = searchQuery.variables.map
+    const selecteds = this.getSelecteds(query, map)
+    const isDisabled = this.countSelecteds(selecteds) === 1
+
+    return FACETS_KEYS.map(key => {
+      if (facets[key]) {
+        if (key === 'SpecificationFilters') {
```
Comment (andreldsa) [inline react/SearchResult.js]: Extract to switch/case and extract the strings to constants.

File: react/SearchResult.js [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    if (!this.props.facetsQuery || !this.props.facetsQuery.facets) return
+
+    const { facetsQuery: { facets }, searchQuery } = this.props
+    const query = searchQuery.variables.query
+    const map = searchQuery.variables.map
+    const selecteds = this.getSelecteds(query, map)
+    const isDisabled = this.countSelecteds(selecteds) === 1
+
+    return FACETS_KEYS.map(key => {
+      if (facets[key]) {
```
Comment (andreldsa) [inline react/SearchResult.js]: User ternary, eg: 

facets[key] ? ... : return

File: react/SearchResult.js [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+  }
+
+  renderSearchFilters() {
+    if (!this.props.facetsQuery || !this.props.facetsQuery.facets) return
+
+    const { facetsQuery: { facets }, searchQuery } = this.props
+    const query = searchQuery.variables.query
+    const map = searchQuery.variables.map
+    const selecteds = this.getSelecteds(query, map)
+    const isDisabled = this.countSelecteds(selecteds) === 1
```
Comment (andreldsa) [inline react/SearchResult.js]: Magic number

File: react/SearchResult.js [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+  getFacetsFromURL,
+  getSearchParamsFromURL,
+  getPagesArgs,
+} from './constants/SearchHelpers'
+import SortOptions from './constants/SortOptions'
+import VTEXClasses from './constants/CSSClasses'
+import { facetsQueryShape, searchQueryShape } from './constants/PropTypes'
+import './global.css'
+
+const FACETS_KEYS = ['Departments', 'CategoriesTrees', 'Brands', 'SpecificationFilters']
```
Comment (andreldsa) [inline react/SearchResult.js]:
```
const FACETS_KEYS = {
  DEPARTMENTS: 'Departments',
  ...
}
```

File: react/SearchResult.js [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+          return facets[key].map(filter => {
+            return (
+              <SearchFilter key={filter.name} title={filter.name}
+                options={filter.facets} type={key} selecteds={selecteds[key]}
+                getLinkProps={this.getLinkProps} disabled={isDisabled} />)
+          })
+        } else if (key === 'CategoriesTrees') {
+          let categories = []
+          facets[key].map(filter => {
+            if (selecteds.Departments.indexOf(filter.Name.toUpperCase()) !== -1) {
```
Comment (andreldsa) [inline react/SearchResult.js]: Magic number

Comment (andreldsa) [reply react/SearchResult.js]: Instead of indexOf comparing if exists, you can use Ramda `find()`.

File: react/SearchResult.js [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+      Departments: [],
+      Brands: [],
+      FullText: [],
+    }
+
+    if (!query && !map) return selecteds
+
+    const pathValues = query.split('/')
+    const mapValues = map.split(',')
+    pathValues.map((val, i) => {
```
Comment (andreldsa) [inline react/SearchResult.js]: Try to use better variables

File: react/SearchResult.js [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+      const valDecoded = decodeURI(val.toUpperCase())
+      if (i > mapValues.length - 1) {
```
Comment (andreldsa) [inline react/SearchResult.js]: lots of magic numbers.

File: react/SearchResult.js [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+        selecteds.Departments.push(valDecoded)
+      } else {
+        if (mapValues[i] === 'c') {
```
Comment (andreldsa) [inline react/SearchResult.js]: Extract to well-named constants.

File: react/SearchResult.js [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+        } else if (mapValues[i] === 'ft') {
+          selecteds.FullText = [valDecoded]
+        }
+      }
+    })
+
+    return selecteds
+  }
+
+  countSelecteds(selecteds) {
```
Comment (andreldsa) [inline react/SearchResult.js]: Use ramda

File: react/SearchResult.js [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+    let count = 0
+    for (const key in selecteds) {
+      if (selecteds[key]) {
+        count += selecteds[key].length
+      }
+    }
+    return count
+  }
+
+  getRecordsFiltered() {
```
Comment (andreldsa) [inline react/SearchResult.js]: Try to use ES6 and upgrade your code style. We use arrow functions in this case.

File: react/components/SearchFilter.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+  constructor(props) {
+    super(props)
+    this.state = {
+      opened: props.opened,
+    }
+  }
+
+  isSelected(optName) {
+    return this.props.selecteds.indexOf(optName.toUpperCase()) !== -1
```
Comment (andreldsa) [inline react/components/SearchFilter.js]: ramda

File: react/components/SelectedFilters.js [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  constructor(props) {
+    super(props)
+    this.state = {
+      opened: props.opened,
+    }
+  }
+
+  getSelectedFilters() {
```
Comment (andreldsa) [inline react/components/SelectedFilters.js]: Ramda

File: react/constants/PropTypes.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
+import PropTypes from 'prop-types'
+
+export const facetOptionShape = PropTypes.shape({
+  Quantity: PropTypes.number.isRequired,
+  Link: PropTypes.string.isRequired,
+  Name: PropTypes.string.isRequired,
+})
+
+export const productShape = PropTypes.shape({
+  productId: PropTypes.string.isRequired,
```
Comment (andreldsa) [inline react/constants/PropTypes.js]: Write the docs to each prop type

File: react/constants/SearchHelpers.js [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+ * e.g: smartphones?map=c
+ * e.g2: eletronics/smartphones?map=c,c
+ */
+export function getFacetsFromURL(pathName, queryParams, isBrand) {
+  const { query, map } = getSearchParamsFromURL(pathName, queryParams, isBrand)
+  return `${query}${map ? `?map=${map}` : ''}`
+}
+
+export function getSearchParamsFromURL(pathName, queryParams, isBrand) {
+  let map = ''
```
Comment (andreldsa) [inline react/constants/SearchHelpers.js]: You can empower this entire file with ramda functions.

File: react/locales/pt-BR.json [hunk 16]
```diff
@@ -1,10 +1,21 @@
...
+  "Categories": "Categorias",
+  "Brands": "Marcas",
+  "Operational System": "Sistema Operacional",
+  "Memory (RAM)": "Memória (Ram)",
+  "search.text": "Exibindo {from}-{to} de {recordsFiltered} resultados",
+  "search.selected-filters": "Filtros Selecionados",
+  "Sales": "Vendas",
+  "Release Date": "Mais recentes",
+  "Discount": "Desconto",
+  "Price, descending": "Mais Caro",
```
Comment (andreldsa) [inline react/locales/pt-BR.json]: Maior Preço

File: react/locales/pt-BR.json [hunk 17]
```diff
@@ -1,10 +1,21 @@
...
+  "Price, ascending": "Mais Barato",
```
Comment (andreldsa) [inline react/locales/pt-BR.json]: Menor Preço

**Comment by andreldsa** [comment]
BTW excellent work 💯

File: react/constants/SearchHelpers.js (hunk 15 above)
Comment (ClaudivanFilho) [reply react/constants/SearchHelpers.js]: how?

File: react/SearchResult.js [hunk 18]
```diff
@@ -22,7 +23,21 @@ import VTEXClasses from './constants/CSSClasses'
 import { facetsQueryShape, searchQueryShape } from './constants/PropTypes'
 import './global.css'
 
-const FACETS_KEYS = ['Departments', 'CategoriesTrees', 'Brands', 'SpecificationFilters']
+const FACETS_KEYS = {
+  Departments: 'Departments',
```
Comment (andreldsa) [inline react/SearchResult.js]: The keys should be typed like constants, DEPARTMENTS, CATEGORIES, etc

File: react/SearchResult.js [hunk 19]
```diff
@@ -109,16 +138,12 @@ class SearchResult extends Component {
...
-    let count = 0
-    for (const key in selecteds) {
-      if (selecteds[key]) {
-        count += selecteds[key].length
-      }
-    }
-    return count
+    // return 0
+    if (!selecteds) return 0
+    return reduce(concat, [])(values(selecteds)).length
```
Comment (andreldsa) [inline react/SearchResult.js]: Really better, try this final touch:

  `countSelecteds = selecteds => return selecteds ? reduce(...).length : 0`

File: react/package.json [hunk 20]
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "vtex.gallery",
-  "version": "1.0.0",
-  "description": "A Gallery Component.",
+  "name": "vtex.search-result",
+  "version": "0.3.0",
```
Comment (andreldsa) [inline react/package.json]: We don't use version.

File: CHANGELOG.md [hunk 21]
```diff
@@ -7,10 +7,4 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
 ## [Unreleased]
 ### Added
-- Internationalization to the schema.
-
-## [0.2.0] - 2018-05-11
-### Added
-- Refactor app to import as default the product-summary app. 
-- **Orderable gallery** Displays a list of products that can be ordered.
-- **Update to the new Pages** Update to new version of Pages.
+- Add initial files
```
Comment (brunojdo) [inline CHANGELOG.md]: ### Renamed 

- App was renamed to `vtex.search-result

File: react/components/SearchHeader.js [hunk 22]
```diff
@@ -0,0 +1,73 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import { FormattedMessage, injectIntl } from 'react-intl'
+
+import Dropdown from '@vtex/styleguide/lib/Dropdown'
```
Comment (brunojdo) [inline react/components/SearchHeader.js]: `import { Dropdown } from 'vtex.styleguide'`

In manifest.json `vtex.styleguide: 4.x`

File: react/components/GalleryItem.js [hunk 23]
```diff
@@ -1,45 +1,18 @@
 import React, { Component } from 'react'
-import PropTypes from 'prop-types'
+
 import ProductSummary from 'vtex.product-summary/ProductSummary'
```
Comment (brunojdo) [inline react/components/GalleryItem.js]: destructure

File: react/components/SearchResultContainer.js [hunk 24]
```diff
@@ -0,0 +1,290 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import { graphql, compose } from 'react-apollo'
+import { reduce, contains, concat, values } from 'ramda'
+
+import ProductSummary from 'vtex.product-summary/ProductSummary'
```
Comment (brunojdo) [inline react/components/SearchResultContainer.js]: idem

File: react/components/SelectedFilters.js [hunk 25]
```diff
@@ -0,0 +1,102 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import { FormattedMessage, injectIntl } from 'react-intl'
+import { Collapse } from 'react-collapse'
+import { reduce, concat, map, keys } from 'ramda'
+
+import Checkbox from '@vtex/styleguide/lib/Checkbox'
```
Comment (brunojdo) [inline react/components/SelectedFilters.js]: idem

File: react/constants/PropTypes.js [hunk 26]
```diff
@@ -0,0 +1,92 @@
+import PropTypes from 'prop-types'
+
+export const facetOptionShape = PropTypes.shape({
```
Comment (brunojdo) [inline react/constants/PropTypes.js]: `propTypes.js`

File: react/locales/en-US.json [hunk 27]
```diff
@@ -1,14 +1,21 @@
...
+  "Categories": "Categories",
+  "Brands": "Brands",
+  "Operational System": "Operational System",
+  "Memory (RAM)": "Memory (Ram)",
+  "search.text": "Showing {from}-{to} from {recordsFiltered} records",
+  "search.selected-filters": "Selected Filters",
+  "Sales": "Sales",
+  "Release Date": "Release Date",
+  "Discount": "Discount",
+  "Price, descending": "Price: High to Low",
```
Comment (brunojdo) [inline react/locales/en-US.json]: `price.descending`
`price.ascending`
`name.ascending`
...

File: react/locales/en-US.json [hunk 28]
```diff
@@ -1,14 +1,21 @@
...
-  "sortBy.lowerPrice": "Price: Low to High",
-  "sortBy.nameAZ": "Name(A-Z)",
-  "sortBy.nameZA": "Name(Z-A)",
-  "sortBy.bestSellers": "Best sellers",
-  "editor.gallery.title": "Gallery",
-  "editor.gallery.description": "A product gallery",
-  "editor.gallery.columnsQuantityLarge.title": "Columns quantity (large viewport)",
-  "editor.gallery.columnsQuantityMedium.title": "Columns quantity (medium viewport)"
-}
+  "Departments": "Departments",
```
Comment (brunojdo) [inline react/locales/en-US.json]: lowercase

Comment (ClaudivanFilho) [reply react/locales/en-US.json]: This key `Departments` comes from server. I do the intl bind directly by the key retrieved from the request.
The same occurs to Categories, Brands, Operational System, Memory (RAM)

File: react/components/SelectedFilters.js (hunk 25 above)
Comment (ClaudivanFilho) [reply react/components/SelectedFilters.js]: In this case is not possible, because there's not Checkbox component in the vtex.styleguide app

**Pull request merged**